### PR TITLE
Implementation of MMC

### PR DESCRIPTION
**Pull request by Callidior**
This PR implements Probabilistic Global Distance Metric Learning (PGDM) mentioned in #13, sometimes also referred to as MMC, originally proposed in the following paper:

> Eric P. Xing, Andrew Y. Ng, Michael I. Jordan and Stuart Russell
> "Distance metric learning with application to clustering with side-information."
> NIPS. Vol. 15. No. 505-512. 2002.

The implementation is mainly a translation of the [Matlab reference code](http://www.cs.cmu.edu/%7Eepxing/papers/Old_papers/code_Metric_online.tar.gz) into Python, only optimized a bit to make use of vectorization and the power of numpy. This makes this implementation 6 times faster than the reference code on the iris dataset.

Both variants of PGDM proposed in the original paper are supported: learning a full or a diagonal Mahalanobis matrix.

The implementation has been validated against the results obtained from the reference code on the iris dataset. Corresponding unit tests are provided.

The class separation score of less than 0.15 on the iris dataset is one of the best scores currently reported in the unit tests for this dataset.

----

Note that PGDM learns only a positive *semi*-definite metric `M`, which precludes obtaining a transformer `L` with `L.T*L=M`  using the Cholesky decomposition. Instead, `PGDM` overrides `transformer()` and uses an eigenvector decomposition `M = V * W * V.T` to obtain a transformer `L = V.T * W^(-1/2)`.

----

**Important note:** PGDM learns a distance function `(x-y)^T * M * (x-y)`, as opposed to `(x-y)^T * inv(M) * (x-y)` as currently stated in the documentation of `metric_learn`. Thus, this PR depends on solving issue #57, e.g., by merging PR #60. If you decide to stick to the current definition of a "metric", `PGDM.metric()` would need to be changed in order to return the inverse of the learned matrix.

### CODE REVIEW

**Comment by Callidior** [comment]
I'm not sure whether PGDM is the correct name for this algorithm. The [survey of Yang](http://www.cs.cmu.edu/%7Eliuy/frame_survey_v2.pdf) introduces this method by Xing et al. as "Supervised Global Distance Metric Learning" and then extends this approach in a subsection to "Probabilistic Global Distance Metric Learning". But the paper and the reference implementation referred to in #13 are not probabilistic at all.

Maybe we should change the name to just "Global Distance Metric Learning", which could be abbreviated as GDML, or to MMC, as suggested by [Weinberger et al.](http://www.jmlr.org/papers/volume10/weinberger09a/weinberger09a.pdf)

How do you think about this?

**Comment by perimosocordiae** [comment]
I agree that the wish list was wrongly conflating this approach with the probabilistic version from the survey, and I'm happy to rename this version to something better.

Despite the referenced usage, I think MMC is a bit of a bad name, as the original paper uses clustering as an example application, not the entire basis for the work. Do any other citations propose nice names for this algorithm? In lieu of any better ideas, GDML can work, even though it's pretty uninformative.

File: test/metric_learn_test.py [hunk 1]
```diff
@@ -149,5 +149,48 @@ def test_iris(self):
...
+    n = self.iris_points.shape[0]
+    a, b, c, d = [], [], [], []
+    for i in range(n):
+      for j in range(i+1, n):
+        if self.iris_labels[i] == self.iris_labels[j]:
+          a.append(i)
+          b.append(j)
+        else:
+          c.append(i)
+          d.append(j)
```
Comment (perimosocordiae) [inline test/metric_learn_test.py]: This can be made a bit nicer with vectorization:
```python
mask = self.iris_labels[None] == self.iris_labels[:,None]
a, b = np.nonzero(np.triu(mask, k=1))
c, d = np.nonzero(np.triu(~mask, k=1))
```

File: test/metric_learn_test.py [hunk 2]
```diff
@@ -149,5 +149,48 @@ def test_iris(self):
...
+
+    # Full metric
+    pgdm = PGDM(convergence_threshold = 0.01)
```
Comment (perimosocordiae) [inline test/metric_learn_test.py]: Style nitpick: don't leave spaces around the `=` in keyword arguments. (here and throughout)

File: metric_learn/pgdm.py [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+import numpy as np
+from six.moves import xrange
+from sklearn.metrics import pairwise_distances
+from sklearn.utils.validation import check_array, check_X_y
+
+from .base_metric import BaseMetricLearner
+from .constraints import Constraints
+
+
+# hack around lack of axis kwarg in older numpy versions
```
Comment (perimosocordiae) [inline metric_learn/pgdm.py]: This code is used in ITML as well, so let's pull this hack out into a new file, `metric_learn/_util.py`.

File: metric_learn/pgdm.py [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+    else:
+      return self._fit_full(X, constraints)
+  
+  def _process_inputs(self, X, constraints):
+    
+    self.X_ = X = check_array(X)
+    
+    # check to make sure that no two constrained vectors are identical
+    a,b,c,d = constraints
+    ident = _vector_norm(X[a] - X[b]) > 1e-9
```
Comment (perimosocordiae) [inline metric_learn/pgdm.py]: In this usage, `ident` is actually a mask where True implies non-identical. Maybe `non_ident` instead?

File: metric_learn/pgdm.py [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+    a, b = a[ident], b[ident]
+    ident = _vector_norm(X[c] - X[d]) > 1e-9
+    c, d = c[ident], d[ident]
```
Comment (perimosocordiae) [inline metric_learn/pgdm.py]: We should also ensure that our constraint vectors aren't empty after this sanity check.

**Comment by Callidior** [comment]
I agree with you that both MMC and GDML are not the most informative names. Unfortunately, I haven't spotted any other term for this algorithm in the existing literature yet and Xing et al. completely avoid giving their algorithm a name (besides just "metric learning").

Although this algorithm is, for sure, applicable to more applications than just clustering, one could also argue that learning a metric that is particularly suitable for clustering is an essential part of the method's objective. It minimizes the total intra-class distance, while keeping the total inter-class distance larger than a given margin. Thus, the classes would be tightly clustered together in the space enforced by such a metric.

However, I leave this decision to you. As soon as you notify me about the final name, I will rename everything and update this PR.

File: metric_learn/pgdm.py [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+    pos_diff = X[a] - X[b]
+    w = np.einsum('ij,ik->jk', pos_diff, pos_diff).ravel()
+    # `w` is the sum of all outer products of the rows in `pos_diff`.
+    # The above `einsum` is equivalent to the much more inefficient:
+    # w = np.apply_along_axis(
+    #         lambda x: np.outer(x,x).ravel(),
+    #         1,
+    #         X[a] - X[b]
+    #     ).sum(axis = 0)
+    t = w.dot(A.ravel() / 100.0)
```
Comment (perimosocordiae) [inline metric_learn/pgdm.py]: Is there an explanation for this division by 100? Combined with the division by 10 in `_prepare_inputs`, there's a lot of magic happening to `A`...

File: metric_learn/pgdm.py [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+main speed-bottleneck.
+Since it has initially been designed for clustering applications, one of the
+implicit assumptions of PGDM is that all classes form a compact set, i.e.,
+follow a unimodal distribution, which restricts the possible use-cases of
+this method. However, it is one of the earliest and a still often cited technique.
+
+Adapted from Matlab code at http://www.cs.cmu.edu/%7Eepxing/papers/Old_papers/code_Metric_online.tar.gz
+"""
+
+from __future__ import print_function, absolute_import
```
Comment (perimosocordiae) [inline metric_learn/pgdm.py]: Importing `division` is a good idea as well.

File: metric_learn/pgdm.py [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+    # The above `einsum` is equivalent to the much more inefficient:
+    # w = np.apply_along_axis(
+    #         lambda x: np.outer(x,x).ravel(),
+    #         1,
+    #         X[a] - X[b]
+    #     ).sum(axis = 0)
+    t = w.dot(A.ravel() / 100.0)
+    
+    w1 = w / np.linalg.norm(w) # make `w` a unit vector
+    t1 = t / np.linalg.norm(w) # distance from origin to `w^T*x=t` plane
```
Comment (perimosocordiae) [inline metric_learn/pgdm.py]: We should compute the norm of `w` once and reuse it.

Comment (perimosocordiae) [inline metric_learn/pgdm.py]: Also, style nitpick: use two spaces between the end of the code and the comment character:

    foo()  # comment

File: metric_learn/pgdm.py [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+          break
+      
+      # third constraint: gradient ascent
+      # max: g(A) >= 1
+      # here we suppose g(A) = fD(A) = \sum_{I,J \in D} sqrt(d_ij' A d_ij)
+      
+      obj_previous = self._fD(X, c, d, A_old) # g(A_old)
+      obj = self._fD(X, c, d, A)              # g(A)
+      
+      if ((obj > obj_previous) or (cycle == 0)) and (satisfy):
```
Comment (perimosocordiae) [inline metric_learn/pgdm.py]: Don't need so many parens here:

    if satisfy and (obj > obj_previous or cycle == 0):

File: metric_learn/pgdm.py [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+        
+        # If projection of 1 and 2 is successful, and such projection
+        # imprives objective function, slightly increase learning rate
```
Comment (perimosocordiae) [inline metric_learn/pgdm.py]: imprives -> improves

File: metric_learn/pgdm.py [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+    
+    Note that d_ij*A*d_ij' = tr(d_ij*A*d_ij') = tr(d_ij'*d_ij*A)
+    so, d(d_ij*A*d_ij')/dA = d_ij'*d_ij
+        df/dA = f'(\sum_{ij \in D} \sqrt{tr(d_ij'*d_ij*A)})
+                * 0.5*(\sum_{ij \in D} (1/sqrt{tr(d_ij'*d_ij*A)})*(d_ij'*d_ij))
+    """
+    dim = X.shape[1]
+    diff = X[c] - X[d]
+    M = np.einsum('ij,ik->ijk', diff, diff) # outer products of all rows in `diff`
+    dist = np.sqrt(np.sum(M * A[None,:,:], axis = (1,2)))
```
Comment (perimosocordiae) [inline metric_learn/pgdm.py]: This line might also benefit from conversion to an `np.einsum` call.

**Comment by perimosocordiae** [comment]
Okay, I can live with MMC.

File: metric_learn/pgdm.py (hunk 4 above)
Comment (Callidior) [inline metric_learn/pgdm.py]: I totally agree. I've copied this part from ITML. We should also change it there.

File: metric_learn/pgdm.py (hunk 6 above)
Comment (Callidior) [inline metric_learn/pgdm.py]: I also don't like this, but that magic is in the original reference code without any comment.

The division by 10 in `_prepare_inputs` could be removed and the learned matrix `A` would then be scaled by a factor of 10. That doesn't hurt the learned metric in any way, but the result wouldn't be identical to that of the reference implementation any more, so I left it this way.

**Comment by Callidior** [comment]
Thanks for your valuable comments!

I hope to find time for implementing the requested changes until next week.

**Comment by Callidior** [comment]
@perimosocordiae I've addressed your change requests and renamed to algorithm to MMC.

File: metric_learn/mmc.py [hunk 12]
```diff
@@ -89,10 +80,14 @@ def _process_inputs(self, X, constraints):
...
-    ident = _vector_norm(X[c] - X[d]) > 1e-9
-    c, d = c[ident], d[ident]
+    no_ident = vector_norm(X[a] - X[b]) > 1e-9
+    a, b = a[no_ident], b[no_ident]
+    no_ident = vector_norm(X[c] - X[d]) > 1e-9
+    c, d = c[no_ident], d[no_ident]
+    if len(a) == 0:
+      raise RuntimeError('No similarity constraints given for MMC.')
+    if len(c) == 0:
+      raise RuntimeError('No dissimilarity constraints given for MMC.')
```
Comment (perimosocordiae) [inline metric_learn/mmc.py]: I think a `ValueError` makes more sense here. We should also specify that no _non-trivial_ constraints were provided, otherwise the user might be confused.

File: metric_learn/mmc.py [hunk 13]
```diff
@@ -308,9 +304,9 @@ def _fD1(self, X, c, d, A):
     """
     dim = X.shape[1]
     diff = X[c] - X[d]
-    M = np.einsum('ij,ik->ijk', diff, diff) # outer products of all rows in `diff`
-    dist = np.sqrt(np.sum(M * A[None,:,:], axis = (1,2)))
-    sum_deri = np.sum(M / (2 * (dist[:,None,None] + 1e-6)), axis = 0)
+    M = np.einsum('ij,ik->ijk', diff, diff)    # outer products of all rows in `diff`
+    dist = np.sqrt(np.einsum('ijk,jk', M, A))  # equivalent to: np.sqrt(np.sum(M * A[None,:,:], axis=(1,2)))
+    sum_deri = np.einsum('ijk,i->jk', M, 0.5 / (dist + 1e-6))  # equivalent to: np.sum(M / (2 * (dist[:,None,None] + 1e-6)), axis=0)
```
Comment (perimosocordiae) [inline metric_learn/mmc.py]: Lines shouldn't go longer than 80 chars, so I'd put the comments on the previous line.

File: metric_learn/mmc.py [hunk 14]
```diff
@@ -344,10 +340,11 @@ def _D_constraint(self, X, c, d, w):
...
     dist = np.sqrt(diff_sq.dot(w))
-    sum_deri1 = np.sum(diff_sq / (2 * np.maximum(dist, 1e-6))[:,None], axis = 0)
-    sum_deri2 = np.sum(
-      np.einsum('ij,ik->ijk', diff_sq, diff_sq) / (-4 * np.maximum(1e-6, dist**3)[:,None,None]),
-      axis = 0
+    sum_deri1 = np.einsum('ij,i', diff_sq, 0.5 / np.maximum(dist, 1e-6))
+    sum_deri2 = np.einsum(
+      'ijk,i',
+      np.einsum('ij,ik->ijk', diff_sq, diff_sq),
+      -0.25 / np.maximum(1e-6, dist**3)
```
Comment (perimosocordiae) [inline metric_learn/mmc.py]: Nested calls to `np.einsum` are pretty hard to read. I think the previous version was fine, but you could also pull out the inner `einsum` to a local variable if this version is significantly faster.

It's also possible to do `einsum` with >2 arrays, but that can have unintended consequences (see https://github.com/numpy/numpy/issues/5366). As of numpy 1.12 the situation is a bit better (see https://github.com/numpy/numpy/pull/5488), but I don't think we need to go down that route.

Comment (Callidior) [inline metric_learn/mmc.py]: Neither do I think so. However, this version is about 4 times faster, since it does not require broadcasting `dist` from 1-D to 3-D. So, I will just pull out the inner `einsum` to a local variable.

Comment (Callidior) [inline metric_learn/mmc.py]: I have found an even 10 times faster way to compute this without intermediate result in a single `einsum` call:

    sum_deri2 = np.einsum(
        'ij,ik->jk',
        diff_sq,
        diff_sq / (-4 * np.maximum(1e-6, dist**3))[:,None]
    )

**Comment by Callidior** [comment]
@perimosocordiae Thanks, fixed :)

**Comment by perimosocordiae** [comment]
Merged. Thanks for the PR, @Callidior!

**Comment by Callidior** [comment]
Thanks for your helpful comments!

But please be aware that the metric learned by this algorithm defines a distance `(x-y)^T * M * (x-y)` and not `(x-y)^T * M^(-1) * (x-y)` as stated in the readme. But some other existing metric learner implementations do the same and hopefully we can get rid of that inconsistency soon :)
